### PR TITLE
Reconfigurator: Add live test that executes expunged zones that were never in service

### PR DESCRIPTION
**Pull request by jgallagher**
This is an attempt to catch other issues like #10025, and implements the reproduction steps described there as a live test, but applied to most zone types and not just Nexus. (We skip multinode clickhouse, because those aren't deployed by default, and internal DNS, because the planner can't replace it without execution running first anyway).

This took a bunch of tries to get passing, and I'd be very unsurprised if there are other kinds of flakes still lurking here. We're not running live tests as a part of CI, so I'm not sure how worried to be about this.

Includes a few bits of live test housekeeping (updates to the README and racklette serial number sets).

### CODE REVIEW

**Comment by jgallagher** [comment]
Running this test on `london` against a branch that did not have the fix for #10025, we see the failure we'd expect: the test times out waiting for blueprint execution to succeed:

```
  stderr ---
    log file: /var/tmp/test_execute_expunged_zone-78ccd669b96cb1ec-test_execute_expunged_zone.11905.0.log
    note: configured to log to "/var/tmp/test_execute_expunged_zone-78ccd669b96cb1ec-test_execute_expunged_zone.11905.0.log"
    note: using DNS from system config (typically /etc/resolv.conf)

    thread 'test_execute_expunged_zone' (2) panicked at live-tests/tests/test_execute_expunged_zone.rs:367:6:
    waited for successful execution: TimedOut(180.373621863s)
    note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

and the log file shows the exact error we saw in #10025:

```
17:21:17.796Z WARN test_execute_expunged_zone: execution had an error
    error = step failed: Ensure external networking resources: Internal Error: unexpected database error: Record not found
```

---

Running against a branch that does have that fix (as well as #10072, which fell out of developing this test), we get a pass but it takes a while; most of this time is waiting for cockroach to be healthy again after expunging one of its nodes:

```
    Starting 1 test across 3 binaries (2 tests skipped)
        SLOW [> 60.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>120.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>180.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>240.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>300.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>360.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>420.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>480.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>540.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>600.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>660.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>720.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>780.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>840.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        SLOW [>900.000s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
        PASS [ 949.973s] omicron-live-tests::test_execute_expunged_zone test_execute_expunged_zone
------------
     Summary [ 949.974s] 1 test run: 1 passed (1 slow), 2 skipped

File: live-tests/tests/common/mod.rs [hunk 1]
```diff
@@ -69,6 +68,23 @@ impl LiveTestContext {
...
+    /// a `DataStore` constructed when this context was created. This method is
+    /// useful if a caller needs to reevaluate what Cockroach instances are
+    /// available in DNS (e.g., due to zone expungement) or needs a `DataStore`
+    /// instance that is not shared.
+    pub async fn new_datastore_connection(
+        &self,
+    ) -> anyhow::Result<(OpContext, Arc<DataStore>)> {
+        let log = &self.logctx.log;
+        let datastore = create_datastore(log, &self.resolver).await?;
+        let opctx = OpContext::for_tests(log.clone(), datastore.clone());
```
Comment (davepacheco) [inline live-tests/tests/common/mod.rs:84]: I was surprised this returned a new `OpContext`, too.  Is that because the older one is still connected to the previous datastore which is still trying to talk to now-missing Cockroach zones?

File: live-tests/README.adoc [hunk 2]
```diff
@@ -22,14 +22,14 @@ First, deploy Omicron using `a4x2` or one of the hardware test rigs.
 
 Ensure the system's target blueprint is enabled. The live tests require this to avoid a case where the live tests generate blueprints based on a target blueprint that is not current, and then make a bunch of changes to the system unrelated to the tests.
 
-On a fresh system, you will have to enable the target blueprint yourself:
```
Comment (davepacheco) [inline live-tests/README.adoc:25]: I take it that you don't need to enable the target blueprint yourself any more?

File: live-tests/tests/common/reconfigurator.rs [hunk 3]
```diff
@@ -78,14 +99,59 @@ pub async fn blueprint_load_target_enabled(
...
+    nexus: &nexus_lockstep_client::Client,
+    edit_fn: F,
+) -> Result<(Blueprint, Blueprint), anyhow::Error>
+where
+    F: FnOnce(&mut BlueprintBuilder) -> Result<(), anyhow::Error>,
+{
+    blueprint_edit_current_target_impl(log, nexus, true, edit_fn).await
+}
+
+/// Modify the system by editing the current target blueprint
```
Comment (davepacheco) [inline live-tests/tests/common/reconfigurator.rs:113]:
```suggestion
/// Modify the system by editing the current target blueprint, verifying that the current target is disabled
```

File: live-tests/tests/common/reconfigurator.rs
Comment (davepacheco) [inline live-tests/tests/common/reconfigurator.rs:80]:
```suggestion
/// Modify the system by editing the current target blueprint, verifying that the current target is enabled
```

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 4]
```diff
@@ -0,0 +1,681 @@
...
+use strum::IntoEnumIterator;
+use update_engine::EventBuffer;
+use update_engine::ExecutionStatus;
+use update_engine::NestedError;
+use update_engine::NestedSpec;
+use update_engine::TerminalKind;
+use update_engine::events::EventReport;
+use update_engine::events::StepOutcome;
+
+#[live_test]
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:45]: Maybe document the scope of what this test is trying to cover?

Maybe something like:

> For each type of zone, test the behavior of executing a blueprint with that zone expunged when we never executed a blueprint with that zone in-service.  This covers edge cases during zone cleanup (where we might try to clean up resources that were never actually set up).

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 5]
```diff
@@ -0,0 +1,681 @@
...
+        }
+        Ok(None) => panic!(
+            "refusing to run live test: no inventory collections exist yet"
+        ),
+        Err(err) => panic!(
+            "refusing to run live test: \
+             error fetching inventory collection: {}",
+            InlineErrorChain::new(&err),
+        ),
+    }
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:73]: I wonder if this should be a helper in `LiveTestContext` (or elsewhere) that gets invoked in all tests.  (Basically, what if some other test runs after this one and runs into this?)

Also, should this be in a `wait_for_condition` to avoid flaking?

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 6]
```diff
@@ -0,0 +1,681 @@
...
+            // no plans to change this in the near future.
+            ZoneKind::ClickhouseKeeper | ZoneKind::ClickhouseServer => {
+                continue;
+            }
+
+            // Skip internal DNS for now - our test relies on the planner
+            // immediately replacing a zone we expunge. The planner (correctly!)
+            // does not do this for internal DNS: it has to wait for inventory
+            // to reflect that the zone has been expunged, which can't happen
+            // during our test because we've disabled execution.
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:87]: Does the planner avoid placing a new internal DNS zone until the old one is definitely gone to avoid its IP address being in two places at once?

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 7]
```diff
@@ -0,0 +1,681 @@
...
+    //
+    // * One zone that had been in-service the last time a blueprint was
+    //   executed is now expunged
+    // * One zone that did not exist the last time a blueprint was executed
+    //   exists in the blueprint but with the expunged disposition
+    // * One zone that did not exist the last time a blueprint was executed
+    //   exists and should be put into service
+
+    // Disable planning and execution.
+    disable_blueprint_planning(nexus, log).await;
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:149]: You updated the README to say that the user should have already done this.  Does that mean we can skip this here?  (Do we also check that it _is_ disabled when we start the live test context?  Feels like we should.)

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 8]
```diff
@@ -0,0 +1,681 @@
...
+                        Err(CondCheckError::NotYet)
+                    } else {
+                        Ok(())
+                    }
+                }
+                Err(err) => Err(CondCheckError::Failed(format!(
+                    "no available event report: {}",
+                    InlineErrorChain::new(&err),
+                ))),
+            }
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:367]: I usually steer people away from depending on specific executions of a background task and instead towards whatever side effect they care about.  In this case I feel like you could wait for inventory collections to reflect that the blueprint zone configs have been propagated to all sleds.  (I think we have a helper for that?)

But I guess you're specifically trying to see if execution generated any problems, even if the configs did get propagated?  It still feels unfortunate you have to do all this work here (though it was built so that you _could_ do this rigorously, so I guess that's something).

Maybe add a comment then about why we're doing it this way and please don't copy/paste it elsewhere?

File: live-tests/tests/test_execute_expunged_zone.rs [hunk 9]
```diff
@@ -0,0 +1,681 @@
...
+                    warn!(
+                        log, "failed to load latest inventory collection";
+                        InlineErrorChain::new(&err),
+                    );
+                    return Err(CondCheckError::NotYet);
+                }
+            };
+
+            // Wait until the inventory collection is sufficiently new.
+            if collection.time_started < wait_start_time {
```
Comment (davepacheco) [inline live-tests/tests/test_execute_expunged_zone.rs:616]: In principle, this could still flake if systems' clocks are not totally in sync.  I'm trying to think if there's some datum in inventory that's more causally related to what you care about.  What you're trying to work out here is that the inventory doesn't predate the expungement you did, right?  Could you use `blueprint_wait_sled_configs_propagated()` here?  I'm thinking if those have been propagated, then the old cockroach isn't running... but I guess you still don't know for sure that cockroach is reporting itself as unhealthy immediately.  This seems tricky.

File: live-tests/tests/common/mod.rs [hunk 10]
```diff
@@ -222,19 +232,19 @@ async fn check_hardware_environment(
...
         "BRM27230037",
         "BRM23230018",
         "BRM23230010",
         // test rig: "berlin"
         "BRM42220011",
         "BRM44220007",
         "BRM42220082",
-        "BRM06240029",
+        "271FVPY0",
     ];
```
Comment (paudmir) [inline live-tests/tests/common/mod.rs:248]: Any chance we can add Dogfood here as well? So that we can run these after an update :)

Comment (jgallagher) [reply live-tests/tests/common/mod.rs:248]: I don't think we should do that - the live tests are considerably more destructive than what I'd be comfortable running on dogfood, because they allow nearly-arbitrary editing of the system's target blueprint.
